source/npm/util.js

```diff
@@ -100,12 +102,70 @@ exports.verifyRecentNpmVersion = async () => {
 };
 
 exports.checkIgnoreStrategy = ({files}) => {
-	const rootDir = pkgDir.sync();
-	const npmignoreExists = fs.existsSync(path.resolve(rootDir, '.npmignore'));
-
-	if (!files && !npmignoreExists) {
+	if (!files && !npmignoreExistsInPackageRootDir()) {
 		console.log(`
 		\n${chalk.bold.yellow('Warning:')} No ${chalk.bold.cyan('files')} field specified in ${chalk.bold.magenta('package.json')} nor is a ${chalk.bold.magenta('.npmignore')} file present. Having one of those will prevent you from accidentally publishing development-specific files along with your package's source code to npm. 
 		`);
 	}
 };
+
+function npmignoreExistsInPackageRootDir() {
+	const rootDir = pkgDir.sync();
+	return fs.existsSync(path.resolve(rootDir, '.npmignore'));
+}
+
+async function getFilesIgnoredByDotnpmignore(fileList) {
+	const whiteList = await ignoreWalker({
+		path: pkgDir.sync(),
+		ignoreFiles: ['.npmignore']
+	});
+	return fileList.filter(minimatch.filter(getIgnoredFilesGlob(whiteList), {matchBase: true}));
+}
+
+function getFilesNotIncludedInTheFilesProperty(globArrayFromFilesProperty, fileList) {
+	const globArrayForFilesAndDirs = [...globArrayFromFilesProperty];
+	const rootDir = pkgDir.sync();
+	for (const glob of globArrayFromFilesProperty) {
+		try {
+			if (fs.statSync(path.resolve(rootDir, glob)).isDirectory()) {
+				globArrayForFilesAndDirs.push(`${glob}/**/*`);
+			}
+		} catch (_) {}
+	}
+
+	return fileList.filter(minimatch.filter(getIgnoredFilesGlob(globArrayForFilesAndDirs), {matchBase: true}));
+}
+
+function getIgnoredFilesGlob(globArrayFromFilesProperty) {
+	/* According to https://docs.npmjs.com/files/package.json#files
+	   npm's default behavior is to ignore these files.
+	   They shouldn't be part of the result returned by function
```

Comment:
This part isn't really necessary (lines 142-143), as lines 140-141 make it clear what this list means and where it comes from.

Reply:
Comment is trimed.